### PR TITLE
Typescript data/ and config/formats

### PR DESCRIPTION
**Pull request by sirDonovan**
Also removes Battle Factory methods accidentally re-added in d0a4a689a721ceb602a3d812a332d3080b5e881e.

Sorry for the large changelog but it was easiest to do everything together. I had to split <code>Format.desc</code> into <code>desc: string</code> and <code>threads: string[]</code> to be able to reuse other interfaces. Otherwise though, the additions are just simple type guards and <code>@ts-ignore</code> everywhere else (until it can be verified that nothing breaks).

### CODE REVIEW

File: data/abilities.js [hunk 1]
```diff
@@ -3041,7 +3056,7 @@ exports.BattleAbilities = {
 	"skilllink": {
 		shortDesc: "This Pokemon's multi-hit attacks always hit the maximum number of times.",
 		onModifyMove: function (move) {
-			if (move.multihit && move.multihit.length) {
+			if (move.multihit && move.multihit instanceof Array && move.multihit.length) {
```
Comment (Slayer95) [inline data/abilities.js]: pls use ``Array.isArray()``

File: data/abilities.js [hunk 2]
```diff
@@ -3342,7 +3357,7 @@ exports.BattleAbilities = {
 		shortDesc: "This Pokemon cannot lose its held item due to another Pokemon's attack.",
 		onTakeItem: function (item, pokemon, source) {
 			if (this.suppressingAttackEvents() && pokemon !== this.activePokemon || !pokemon.hp || pokemon.item === 'stickybarb') return;
-			if ((source && source !== pokemon) || this.activeMove.id === 'knockoff') {
+			if ((source && source !== pokemon) || (this.activeMove && this.activeMove.id === 'knockoff')) {
```
Comment (Slayer95) [inline data/abilities.js]: Maybe it would be better to ``assert(this.activeMove)`` ?

File: data/abilities.js [hunk 3]
```diff
@@ -309,8 +310,8 @@ exports.BattleAbilities = {
 		shortDesc: "Prevents other Pokemon from lowering this Pokemon's Defense stat stage.",
 		onBoost: function (boost, target, source, effect) {
 			if (source && target === source) return;
-			if (boost['def'] && boost['def'] < 0) {
-				delete boost['def'];
+			if (boost.def && boost.def < 0) {
+				delete boost.def;
```
Comment (Zarel) [inline data/abilities.js]: Are these changes actually necessary?

File: data/moves.js [hunk 4]
```diff
@@ -9494,6 +9508,7 @@ exports.BattleMovedex = {
 			onStart: function (side, source) {
 				this.debug('Lunar Dance started on ' + side.name);
 				this.effectData.positions = [];
+				// @ts-ignore
 				for (let i = 0; i < side.active.length; i++) {
```
Comment (Zarel) [inline data/moves.js]: `for (const i of side.active.keys()) {`

File: data/moves.js [hunk 5]
```diff
@@ -9930,7 +9946,7 @@ exports.BattleMovedex = {
 					'chatter', 'counter', 'covet', 'focuspunch', 'mefirst', 'metalburst', 'mirrorcoat', 'struggle', 'thief',
 				];
 				let move = this.getMoveCopy(action.move.id);
-				if (move.category !== 'Status' && !noMeFirst.includes(move)) {
+				if (move.category !== 'Status' && !noMeFirst.includes(move.id)) {
```
Comment (Zarel) [inline data/moves.js]: ! TypeScript caught another bug!

File: data/moves.js [hunk 6]
```diff
@@ -10484,6 +10501,7 @@ exports.BattleMovedex = {
 			},
 			onDamagePriority: -101,
 			onDamage: function (damage, target, source, effect) {
```
Comment (Zarel) [inline data/moves.js]: Out of curiosity, how did you get away with typing these parameters? I've been trying to do that for ages because TypeScripting chat commands is hella tedious without it...

Comment (Zarel) [reply data/moves.js]: Oh, I see, `interface` in `globals.ts`... I should have thought of that...

**Comment by Zarel** [comment]
I massively approve this!

File: data/abilities.js (hunk 2 above)
Comment (sirDonovan) [reply data/abilities.js]: I wouldn't mind but <code>assert()</code> currently isn't used anywhere outside of the test suite. @Zarel which would you prefer?

File: data/abilities.js (hunk 3 above)
Comment (sirDonovan) [reply data/abilities.js]: Yeah, without them you get the <code>Object is possibly 'undefined'.</code> error.

File: config/formats.js [hunk 7]
```diff
@@ -3,7 +3,8 @@
 // Note: This is the list of formats
 // The rules that formats use are stored in data/rulesets.js
 
-exports.Formats = [
```
[the remark this answers is not in the document]
Comment (sirDonovan) [reply config/formats.js]: Yes, without this change it doesn't use the <code>/**@type {}*/</code>.

File: tsconfig.json [hunk 8]
```diff
@@ -11,6 +11,19 @@
...
+        "./data/formats-data.js",
+        "./data/items.js",
+        "./data/learnsets.js",
+        "./data/moves.js",
+        "./data/pokedex.js",
+        "./data/random-teams.js",
+        "./data/rulesets.js",
+        "./data/scripts.js",
+        "./data/statuses.js",
+        "./data/typechart.js",
```
[the remark this answers is not in the document]
Comment (sirDonovan) [reply tsconfig.json]: Whoops, forgot about that. It turns out the <code>.json</code> files don't cause any issues anyway.

File: data/abilities.js (hunk 2 above)
Comment (Zarel) [reply data/abilities.js]: PS actually does assert all over the codebase, it's just that we don't use 

    assert(shouldBeTrue);

we use

    if (!shouldBeTrue) throw new Error("shouldBeTrue isn't true");

The latter _is_ used all over the PS codebase, and is probably the best choice here, assuming you actually want to assert here.

File: data/abilities.js (hunk 3 above)
Comment (Zarel) [reply data/abilities.js]: Hm, can you try again with TypeScript Nightly? It should be able to equivocate the two forms...

File: config/formats.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Zarel) [reply config/formats.js]: Can you use a cast instead? `exports.Formats = /**@type {blah} */ ([`

File: data/abilities.js (hunk 3 above)
Comment (sirDonovan) [reply data/abilities.js]: I'm getting the same error with TypeScript Nightly.

File: config/formats.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (sirDonovan) [reply config/formats.js]: That doesn't work either.

File: data/abilities.js (hunk 3 above)
Comment (Zarel) [reply data/abilities.js]: :[

I guess this can be here. Remind me to screw around with it myself later.

File: config/formats.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Zarel) [reply config/formats.js]: lol come on TypeScript

I guess this works.

[the remark this answers is not in the document]
Comment (Slayer95) [reply config/formats.js]: cc @Antar1011 for ``formats.js`` change.

File: sim/pokemon.js [hunk 9]
```diff
@@ -1251,13 +1253,13 @@ class Pokemon {
...
-	 * @param {Pokemon} [source]
+	 * @param {string | Ability} ability
+	 * @param {?Pokemon} [source]
 	 * @param {boolean} [isFromFormeChange]
 	 */
-	setAbility(abilityName, source, isFromFormeChange) {
+	setAbility(ability, source, isFromFormeChange) {
 		if (!this.hp) return false;
-		let ability = this.battle.getAbility(abilityName);
+		if (typeof ability === 'string') ability = this.battle.getAbility(ability);
```
Comment (Slayer95) [inline sim/pokemon.js]: Why is this needed?

Comment (sirDonovan) [reply sim/pokemon.js]: Not exactly needed, but it's preventing an unnecessary function call if an <code>Ability</code> is passed in.

Comment (Slayer95) [reply sim/pokemon.js]: This seems like premature optimization... Surveying for ``.setAbility()`` calls, an ``Ability`` instance is passed almost nowhere (only in Trace for a couple of old gen mods).

Will TS still know that ``ability`` is always an ``Ability`` instance from next line onwards with this change?

Comment (sirDonovan) [reply sim/pokemon.js]: Yes it will.

Comment (Zarel) [reply sim/pokemon.js]: Yeah, I think the term is "contextual typing" and TypeScript does it and it's amazing and basically the thing about TypeScript I love most. I only wish it were easier to assert a contextual type without providing an explicit guard.

File: chat-plugins/info.js [hunk 10]
```diff
@@ -1482,14 +1482,16 @@ exports.commands = {
 			}
 			let formatType = (format.gameType || "singles");
 			formatType = formatType.charAt(0).toUpperCase() + formatType.slice(1).toLowerCase();
-			if (!format.desc) {
+			if (!format.desc && !format.threads) {
```
Comment (Zarel) [inline chat-plugins/info.js]: Why did we split `.desc` into `.desc` and `.threads`, anyway?

**Comment by Zarel** [comment]
This is definitely worth merging now and figuring out minor problems later, though.

File: chat-plugins/info.js (hunk 10 above)
Comment (sirDonovan) [reply chat-plugins/info.js]: Without the split, you get this error in <code>globals.ts:</code> <code>Interface 'Format' cannot simultaneously extend types 'Effect' and 'FormatsData'.
  Named property 'desc' of types 'Effect' and 'FormatsData' are not identical.</code>

Comment (Zarel) [reply chat-plugins/info.js]: Oh, makes sense.
